### DISCUSSION TITLE
DataGrid 'ServerMode' ItemSource binding

### DISCUSSION THREAD

**Discussion by CzBuCHi** [AvaloniaUI/Avalonia#12004] [Q&A]
i would like to bind DataGrid to `VirtualList` instance that is downloading items from server on demand:

```cs

interface IVirtualListService {
    // query is json that holds DataGrid filters, ordering, grouping, visible columns, etc.  
    // will be used in GetCountAsync and GetItemsAsync so server can generate more specific SQL query
    void SetQuery(object query);

    // uses HttpClient.SendAsync under the hood
    Task<int> GetCountAsync(string dataSource); 
    Task<T[]> GetItemsAsync<T>(string dataSource, int start, int count, string query);
}

class VirtualList<T> : IReadOnlyList<T> {
    private const int PageSize = 128;

    private readonly IVirtualListService _Service;

    // identifier used to determine database table / view on server
    private string _DataSource;

    // local cache for already downlaoded items
    // TODO: remove items, that where changed on server so they can be redownloaded
    private readonly Dictionary<int, T>  _Local = new();

    public IEnumerator<T> GetEnumerator() {
        return new VirtualListEnumerator(this);
    }

    IEnumerator IEnumerable.GetEnumerator() {
        return GetEnumerator();
    }
   
    VirtualList(string dataSource, IVirtualListService service) {
        _DataSource = dataSource;
        _Service = service;
        Count = _Service.GetCountAsync(_DataSource).GetAwaiter().GetResult(); // may return LARGE value
    }

    public int Count { get; }
    
    public T this[int index] {
        get {
            if (!_Local.TryGetValue(index, out var value)) {
                LoadPage((index / PageSize) * PageSize);
                value = _Local[index];
            }
            return value;
        }
    }

    private void LoadPage(int start){
        var items = _Service.GetItemsAsync<T>(_DataSource, start, _PageSize).GetAwaiter().GetResult(); 
        for(var i = start, j = 0; i < PageSize && j < items.Length; ++i, ++j){
            _Local[i] = items[j];
        }
    }
    
    private class VirtualListEnumerator : IEnumerator<T>
    {
        private readonly VirtualList<T> _VirtualList;

        private int _Index = -1;
        public VirtualListEnumerator(VirtualList<T> virtualList) {
            _VirtualList = virtualList;
        }

        public bool MoveNext() {
            return ++_Index < _VirtualList.Count;
        }

        public void Reset() {
            _Index = -1;
        }

        public T Current => _VirtualList[_Index];

        object? IEnumerator.Current => Current;

        public void Dispose() {}
    }
}
```

currently when i bind `ItemsSource` property on datagrid will simply call `GetEnumerator` ( == download ALL rows from server) and im not sure how to tell it to use Count + indexer instead ...

Ideally grid would behave like this:

1. detects that its ItemSource is 'virtual' (probably by checking if ItemSource implements some marker interface), if not then use GetEnumerator to populate data (aka like now)
2. reads `ItemSource.Count` property
3. generate rows with theyr datacontexts set to something like `new RowIndex(/*row index in grid*/)` 
4. when row is rendered it will check its datacontext and if set to `RowIndex` it will replace it with value from `Grid.ItemSource[RowIndex.Index]`
5. RowIndex could have special datatemplate, that will render empty row as placeholder to prevent flickering

in the end if user opens grid, that is showing say 20 rows then only first PageSize items will be downloaded from server and not entire set ...

**Comment by jp2masa** [comment]
You're not implementing `IReadOnlyList<T>` correctly. `Count` and `GetEnumerator` must match. So, you have two options:

- Only return the downloaded items in `Count` and `GetEnumerable`, and raise property changed as needed.
- Return full count in `Count` and return all items in `GetEnumerable`, each wrapped in a viewmodel responsible for tracking its own download state.

**Reply by CzBuCHi** [reply]
im not sure how my implementation of  `Count` and `GetEnumerator` do not match ...
only way i can think of is when someone insert/delete data on server after count request is made ... (then getitem request may use index out of range), i need to rethink how to handle this ...

ad1) DataGrid needs to know total Count, so it can correctly render scrollbar ...
ad2) i think i already do something like that - except my tracking code is in VirtualList indexer

**Reply by jp2masa** [reply]
You're right, I initially thought they didn't match and forgot to update my response. The options I listed are still valid, though.

Given that the 2nd option is the one that fits your use case, you just have to wrap each item in a viewmodel that tracks the async operation. You start with placeholder values in the item viewmodel, then when the task completes (success or failure), you update the values accordingly (make sure to update the values on the UI thread).

**Reply by CzBuCHi** [reply]
my primary reason for this is download only what is actually needed and not whole set (i cannot imagine situation, where i would need to download and show all 10_000_000 rows from database ... ) so there will be no task, that is downloading given row until actually needed ...

anyway im still thinking, this should / could have direct support in DataGrid (basically something like [DevExpress Grid 'ServerMode'](https://docs.devexpress.com/WPF/6279/controls-and-libraries/data-grid/bind-to-data/server-mode-and-instant-feedback-mode)) so users dont need to reinvent the wheel for 4418498 time ...

**Reply by jp2masa** [reply]
I'm not sure if DataGrid supports virtualization. If it doesn't, I think TreeDataGrid does (https://github.com/AvaloniaUI/Avalonia.Controls.TreeDataGrid).

Assuming virtualization, I'm also not sure if it supports `IReadOnlyList<T>`, did you try implementing `IList<T>`?
